core/arch/arm/kernel/thread.c

```diff
@@ -1344,34 +1363,88 @@ static void thread_rpc_alloc(size_t size, size_t align, unsigned int bt,
 
 	reg_pair_from_64(carg, rpc_args + 1, rpc_args + 2);
 	thread_rpc(rpc_args);
+
 	if (arg->ret != TEE_SUCCESS)
 		goto fail;
 
 	if (arg->num_params != 1)
 		goto fail;
 
-	if (params[0].attr != OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT)
-		goto fail;
+	if (params[0].attr == OPTEE_MSG_ATTR_TYPE_NESTED) {
+		paddr_t *pages;
+		paddr_t page_offset;
+		int num_pages;
+		struct mobj *nested_mobj;
+		struct optee_msg_param *nested_params;
+
+		*cookie = params[0].u.nested.shm_ref;
+		page_offset = params[0].u.nested.buf_ptr & SMALL_PAGE_MASK;
+
+		nested_mobj = map_params_buffer(params[0].u.nested.buf_ptr,
+						params[0].u.nested.params_num);
+		if (!nested_mobj)
+			goto free_first;
+
+		nested_params = mobj_get_va(nested_mobj, page_offset);
+		if ((nested_params[0].attr & OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT)
```

Comment:
OK, please add a `msg_param_attr_is_tmem()` function and use that here instead for clarity. A comment that this isn't a security check would also be helpful.